### PR TITLE
FIX: Evaluation of ??

### LINKED ISSUE #1623: Nullish Coalescing operator - functions shouldn't be evaluated multiple times
Spec clearly says to evaluate the CoalesceExpression _once_:

https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#prod-CoalesceExpression

[CoalesceExpression](https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#prod-CoalesceExpression) : [CoalesceExpressionHead](https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#prod-CoalesceExpressionHead) ?? [BitwiseORExpression](https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#prod-BitwiseORExpression)
1. Let lRef be ? [Evaluation](https://tc39.es/ecma262/multipage/syntax-directed-operations.html#sec-evaluation) of [CoalesceExpressionHead](https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#prod-CoalesceExpressionHead).
2. Let lVal be ? [GetValue](https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#sec-getvalue)(lRef).
3. If lVal is either undefined or null, then
a. Let rRef be ? [Evaluation](https://tc39.es/ecma262/multipage/syntax-directed-operations.html#sec-evaluation) of [BitwiseORExpression](https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#prod-BitwiseORExpression).
b. Return ? [GetValue](https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#sec-getvalue)(rRef).
4. Else,
a. Return lVal.

Yet, we generate tree like this, that evaluates it multiple times.

```
    VAR 2
        NAME eval1 2
            HOOK
                OR
                    SHEQ
                        NULL
                        CALL 2
                            NAME f 2
                    SHEQ
                        NAME undefined
                        CALL 2
                            NAME f 2
                NUMBER 42.0 2
                CALL 2
                    NAME f 2
```

 I'm working on fixing this. (Thanks @andreabergia for noticing!)

### PR DESCRIPTION
**Pull request by 0xe**
Fixes #1623

### CODE REVIEW

File: tests/src/test/java/org/mozilla/javascript/tests/NullishCoalescingOpTest.java [hunk 1]
```diff
@@ -56,4 +56,23 @@ public void testNullishColascingPrecedence() {
...
+
+    @Test
+    public void testNullishColascingEvalOnce() {
+        Utils.runWithAllOptimizationLevels(
+                cx -> {
+                    Scriptable scope = cx.initStandardObjects();
+                    cx.setLanguageVersion(Context.VERSION_ES6);
+
+                    String script1 =
+                            "var runce = 0; \n"
```
Comment (p-bakker) [inline tests/src/test/java/org/mozilla/javascript/tests/NullishCoalescingOpTest.java:68]: runs or count instead of runce

**Comment by p-bakker** [comment]
Assume this isn't covered by any test from test262?

**Comment by 0xe** [comment]
> Assume this isn't covered by any test from test262?

Yeah, this isn't covered. Will raise a PR against 262 to add this scenario.

**Comment by 0xe** [comment]
> Looks like the pipeline fails due to some failing tests

I'm working on fixing those.

File: rhino/src/main/java/org/mozilla/javascript/IRFactory.java [hunk 2]
```diff
@@ -758,7 +758,37 @@ private Node transformIf(IfStatement n) {
     private Node transformInfix(InfixExpression node) {
         Node left = transform(node.getLeft());
         Node right = transform(node.getRight());
-        return createBinary(node.getType(), left, right);
+        if (node.getType() == Token.NULLISH_COALESCING) {
```
Comment (p-bakker) [inline rhino/src/main/java/org/mozilla/javascript/IRFactory.java:761]: Maybe ternary operator instead of if/else?

**Comment by gbrail** [comment]
Looks good to me too. Thanks!

**Pull request merged**